core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java

```diff
@@ -205,4 +205,29 @@ private ShardRouting initializingShard(ShardRouting shardRouting, String current
         assert initializingShard.initializing();
         return initializingShard;
     }
+
+    @Override
+    public Decision canRemainOnNode(RoutingNode node, RoutingAllocation allocation) {
+        int primaryNodeOutRecoveries = allocation.routingNodes().getOutgoingRecoveries(node.nodeId());
+        if (primaryNodeOutRecoveries >= concurrentOutgoingRecoveries) {
+            return allocation.decision(THROTTLE, NAME, "too many outgoing shards are currently recovering [%d], limit: [%d]",
+            primaryNodeOutRecoveries, concurrentOutgoingRecoveries);
+        } else {
+            return allocation.decision(YES, NAME, "below shard recovery limit of outgoing: [%d < %d]", primaryNodeOutRecoveries,
+            concurrentOutgoingRecoveries);
+        }
+    }
+    
+    @Override
+    public Decision canMoveAnyShardFromNode(RoutingNode node, RoutingAllocation allocation) {
+        int outgoingRecoveries = allocation.routingNodes().getOutgoingRecoveries(node.nodeId());
```

Comment:
This is fundamentally broken I think. If we return NO here, we will not try to move shards away from this node. This is only correct for primary shards on this node, however, as moving a replica shard away from this node will not count as outgoing recovery for this node (but will count as outgoing recovery for the node that has the corresponding primary).

Reply:
During a relocation when there are large number of shards in the cluster it wouldn't matter as the node which we are skipping already has sufficient outgoing recoveries although it could have relocated a replica in the example below which could have caused an outgoing recovery on the node hosting the peer primary of the shard. As a consequence the node having the peer primary will choose to now relocate one of it's primary shard to hit the concurrent outgoing recovery(instead of a peer replica recovery). Lets take an example where we have to relocate shards from 2 nodes(node3,4) to 4 nodes (node 1,2,5,6). With the current proposal node3 will be removed from the nodeIteration once [test][3] P has two recoveries one for the replica and the other for the primary getting relocated to node2 while it could have accomodated moving a replica as well. But this seems to have been covered by node4 which relocates its 2 primary shards
````
Default node concurrent recoveries: 2
Shard        Source      State     Target   P/R 
[test][3], node[node4], relocating [node5], [R], 
[test][2], node[node4], relocating [node6], [P], 
[test][0], node[node4], relocating [node1], [P], 
[test][3], node[node3], relocating [node2], [P], 
````
In the above case the existing algorithm would have given a fare share to both the nodes,as in a replica from node3 would have relocated whose primary lies in node2 giving it a form like

````
Default node concurrent recoveries: 2
Shard        Source      State     Target   P/R 
[test][3], node[node4], relocating [node5], [R], 
[test][2], node[node4], relocating [node6], [P], 
[test][2], node[node3], relocating [node1], [R], 
[test][3], node[node3], relocating [node2], [P], 
````
When the node doesn't have sufficient shards to hit the outgoing recoveries the node will not be skipped ensuring that all shards gets eventually relocated. So the logic doesn't break any existing functionality it just compensates some primary shards with replica shards

Reply:
> During a relocation when there are large number of shards in the cluster it wouldn't matter

While in your situation this won't be an issue, it will be observable in smaller clusters. If the allocation deciders don't respect the contract they set out to fulfil, it will be even harder to understand shard movements. The allocation explain API would also have to become aware of this oddity and expose it to the end user.

While not as optimal (because pruning cannot happen at the node level), might it be good enough to have a method `canMoveAway(ShardRouting shardRouting, RoutingAllocation allocation)`? This would then be implemented by `ThrottlingAllocationDecider` doing the outgoing recoveries check, taking the node of the primary shard into account.